pkg/sql/schemachanger/sctest/end_to_end.go

```diff
@@ -127,26 +127,33 @@ func EndToEndSideEffects(t *testing.T, relPath string, newCluster NewClusterFunc
 	numTestStatementsObserved := 0
 	var setupStmts parser.Statements
 	datadriven.RunTest(t, path, func(t *testing.T, d *datadriven.TestData) string {
-		sqlutils.VerifyStatementPrettyRoundtrip(t, d.Input)
-		stmts, err := parser.Parse(d.Input)
-		require.NoError(t, err)
-		require.NotEmpty(t, stmts)
-		execStmts := func() {
-			for _, stmt := range stmts {
-				tdb.Exec(t, stmt.SQL)
+		parseStmts := func() (parser.Statements, func()) {
+			sqlutils.VerifyStatementPrettyRoundtrip(t, d.Input)
+			stmts, err := parser.Parse(d.Input)
+			require.NoError(t, err)
+			require.NotEmpty(t, stmts)
+			return stmts, func() {
+				for _, stmt := range stmts {
+					tdb.Exec(t, stmt.SQL)
+				}
+				waitForSchemaChangesToSucceed(t, tdb)
 			}
-			waitForSchemaChangesToSucceed(t, tdb)
 		}
-
 		switch d.Cmd {
 		case "setup":
+			stmts, execStmts := parseStmts()
 			a := prettyNamespaceDump(t, tdb)
 			execStmts()
 			b := prettyNamespaceDump(t, tdb)
 			setupStmts = stmts
 			return sctestutils.Diff(a, b, sctestutils.DiffArgs{CompactLevel: 1})
-
+		case "stage-exec":
+			fallthrough
+		case "stage-query":
+			// Nothing to do in this case, expect the expected output.
+			return d.Expected
```

Comment:
How did this work? 
1. If it's a `stage-exec` command, it "falls through" so we enter the next case (case "stage-query") and we return d.Expected there. Where do we actually execute the stmt specified in a `stage-exec` command?
2. In addition to 1, where is the code that reads and acts upon various "stage-exec/query configurations" (i.e. run at this stage in this phase)

Reply:
This is for the end-to-end variant, which will never act on these statements. The cumulative.go has the parser for actually executing statements, the two tests are different since one does dependency injection and tracks side effects. Whereas the other variants actually go through normal runtime code paths (which is the same version for testing backups, rollbacks, etc..).

Reply:
I added a comment to make this clear in both places.